src/main/java/com/levik/pool/CustomConnectionDataSource.java

```diff
@@ -0,0 +1,96 @@
+package com.levik.pool;
+
+import lombok.SneakyThrows;
+import org.postgresql.ds.PGSimpleDataSource;
+
+import javax.sql.DataSource;
+import java.io.PrintWriter;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.logging.Logger;
+
+public class CustomConnectionDataSource implements DataSource {
+
+    private static final int DEFAULT_CONNECTION_POOL_SIZE = 10;
+    private final Queue<ConnectionWrapper> connectionPool = new ArrayDeque<>();
+
+    private final String url;
+    private final String username;
+    private final String password;
+
+    @SneakyThrows
+    public CustomConnectionDataSource(String url, String username, String password) {
+        this.url = url;
+        this.username = username;
+        this.password = password;
+
+        createConnectionPool(DEFAULT_CONNECTION_POOL_SIZE);
+    }
+
+    @SneakyThrows
+    public CustomConnectionDataSource(String url, String username, String password, int poolSize) {
+        this.url = url;
+        this.username = username;
+        this.password = password;
+
+        createConnectionPool(poolSize);
+    }
+
+    @SneakyThrows
+    private void createConnectionPool(int poolSize) {
+        for (int i = 0; i < poolSize; i++) {
+            var connection = DriverManager.getConnection(url, username, password);
```

Comment:
I think we can driver once before the loop and open a connection with Driver.class, it is not critical but maybe save few seconds)